app/src/main/java/org/olpc/xorobothash/MainActivity.java

```diff
@@ -0,0 +1,157 @@
+package org.olpc.xorobothash;
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+import android.os.AsyncTask;
+import android.support.v7.app.ActionBarActivity;
+import android.os.Bundle;
+import android.view.Menu;
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.ImageView;
+import android.util.Log;
+import android.widget.TextView;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.net.HttpURLConnection;
+
+public class MainActivity extends ActionBarActivity {
+
+    private ImageView robotImageView;
+
+    //For debugging
+    private static final String TAG = "XORobotHash";
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+        robotImageView = (ImageView) findViewById(R.id.robot_hash_image);
+        getRobot();
+    }
+
+    private void getRobot() {
+        boolean isConnected = checkNetworkConnectivity();
+        boolean done = updateView(isConnected);
+        if (!done) {
+           if(isConnected) {
+               // Will call updateView again when done
+               new DownloadRobotImage().execute(uniqueRoboHashURL(), robotImagePath());
+           }
+           // Nothing else to do, updateView already showed the right error message
+        }
+    }
+
+    private boolean updateView(boolean isConnected) {
+        Log.i(TAG, "In updateView()");
+        boolean robotVisible = false;
+        File file = new File(robotImagePath());
+        // Check if we have ever successfully downloaded the XO's robot image
+        if (file.exists()) {
+            Log.i(TAG, "Robohash image file exists.");
+            robotVisible = showRobot(robotImagePath());
```

Comment:
Avoid calling `robotImagePath` twice by passing the `file` to `showRobot`, rather than the path string. You can get the string back out with `File.getAbsolutePath()` later on.